starlette/staticfiles.py

```diff
@@ -66,5 +97,7 @@ def __init__(self, scope: Scope, path: str, check_directory: str = None) -> None
                 response = FileResponse(
                     self.path, stat_result=stat_result, method=self.scope["method"]
                 )
+                if self.is_not_modified(response):
```

Comment:
This looks like a strange way around to do this. Could we not do something more like:

```python
if not stat.S_ISREG(mode):
    response = PlainTextResponse("Not Found", status_code=404)
elif self.is_not_modified(stat_result):
    response = ...
else:
    response = FileResponse(...)
```

Reply:
Hi, I just realized that you need `etag` and `if-modified-since` header for checking 304. So I move the `set_stat_header()` function to classmethod. What do you think?

Maybe we can learn `whitenoise` here, in `StaticFile` they got the `etag` and `if-modified-since` in the first place:

https://github.com/evansd/whitenoise/blob/1b5a7f3a0f151ee1449d88da9e4813644e435fd4/whitenoise/responders.py#L36-L38